### PR TITLE
8266950: Remove vestigial support for non-strict floating-point execution

### PR DESCRIPTION
**Pull request by dholmes-ora**
As part of JEP 306, the vestiges of HotSpot support for non-strict floating-point execution can be removed. All methods implicitly have strictfp semantics so the explicit checks for is_strict() can be replaced by true and the code reformulated accordingly.

There are still some names that include "strict" that could potentially be renamed to remove it, but the fact we have to have strict fp semantics is still important on some platforms, so the names help reinforce that IMO.

Testing: tiers 1-3

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266950](https://bugs.openjdk.java.net/browse/JDK-8266950): Remove vestigial support for non-strict floating-point execution

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 4dcab9b7d5517af19a84771b1dcb035fd4d5302d
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to c0c35a7774f63c9ad1e17419d3aa6f5eec7eb928

### Contributors
 * Vladimir Ivanov `<vlivanov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3991/head:pull/3991` \
`$ git checkout pull/3991`

Update a local copy of the PR: \
`$ git checkout pull/3991` \
`$ git pull https://git.openjdk.java.net/jdk pull/3991/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3991`

View PR using the GUI difftool: \
`$ git pr show -t 3991`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3991.diff">https://git.openjdk.java.net/jdk/pull/3991.diff</a>

</details>

### CODE REVIEW

**Review by iwanowww — APPROVED** [review]
Overall, it looks very good.
Thanks for taking care of compiler part, David.

I think it makes sense to remove lir_div_strictfp and lir_mul_strictfp in C1 as well:
  https://github.com/openjdk/jdk/pull/4027

Feel free to incorporate the patch into the current PR if you agree with the change.
(Passed hs-tier1 - hs-tier4 testing and x86_32 build.)

Otherwise, I'll handle it as a separate PR.

**Comment by vnkozlov** [comment]
@iwanowww I agree with your suggestion but lets do it in separate RFE.

**Review by vnkozlov — APPROVED** [review]

**Comment by dholmes-ora** [comment]
@iwanowww  and @vnkozlov  I have merged @iwanowww 's changes with this PR and am re-testing.

Thanks,
David

**Comment by dholmes-ora** [comment]
Can I please get re-reviews for this.

Thanks,
David

**Review by iwanowww — APPROVED** [review]

**Comment by iwanowww** [comment]
There are some suspicious failures on linux-x86 in pre-submit testing results:
- compiler/c1/Test6855215.java
- compiler/intrinsics/string/TestStringLatin1IndexOfChar.java

The tests explicitly specify `-XX:UseSSE=0`, so it may be related to the patch. Anybody interested in linux-x86 want to take a look? @shade @DamonFool

**Review by vnkozlov — APPROVED** [review]

**Comment by DamonFool** [comment]
> There are some suspicious failures on linux-x86 in pre-submit testing results:
> 
> * compiler/c1/Test6855215.java
> * compiler/intrinsics/string/TestStringLatin1IndexOfChar.java
> 
> The tests explicitly specify `-XX:UseSSE=0`, so it may be related to the patch. Anybody interested in linux-x86 want to take a look? @shade @DamonFool

compiler/c1/Test6855215.java crashes on x86_32 with the latest patch.
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/jdk/src/hotspot/cpu/x86/c1_LinearScan_x86.cpp:683), pid=13254, tid=13270
#  assert(op2->tmp1_opr()->is_fpu_register()) failed: strict operations need temporary fpu stack slot
#
# JRE version: OpenJDK Runtime Environment (17.0) (fastdebug build 17-internal+0-adhoc..jdk)
# Java VM: OpenJDK Server VM (fastdebug 17-internal+0-adhoc..jdk, mixed mode, sharing, tiered, g1 gc, linux-x86)
# Problematic frame:
# V  [libjvm.so+0x64fcf0]  FpuStackAllocator::handle_op2(LIR_Op2*)+0x120
#

Current CompileTask:
C1:    109   66    b  3       java.util.HashMap::resize (356 bytes)

Stack: [0xa1f7f000,0xa2000000],  sp=0xa1ffe540,  free space=509k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x64fcf0]  FpuStackAllocator::handle_op2(LIR_Op2*)+0x120
V  [libjvm.so+0x65232c]  FpuStackAllocator::allocate_block(BlockBegin*)+0x28c
V  [libjvm.so+0x6526af]  FpuStackAllocator::allocate()+0xcf
V  [libjvm.so+0x652a4d]  LinearScan::allocate_fpu_stack()+0xcd
V  [libjvm.so+0x648d7a]  LinearScan::do_linear_scan()+0x39a
V  [libjvm.so+0x588f40]  Compilation::emit_lir()+0xd10
V  [libjvm.so+0x58ba60]  Compilation::compile_java_method()+0x720
V  [libjvm.so+0x58c287]  Compilation::compile_method()+0x247
V  [libjvm.so+0x58cbe3]  Compilation::Compilation(AbstractCompiler*, ciEnv*, ciMethod*, int, BufferBlob*, bool, DirectiveSet*)+0x3d3
V  [libjvm.so+0x58de72]  Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1b2
V  [libjvm.so+0x83f5fb]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xc6b
V  [libjvm.so+0x83ff71]  CompileBroker::compiler_thread_loop()+0x491
V  [libjvm.so+0x86b8c6]  CompilerThread::thread_entry(JavaThread*, JavaThread*)+0x56
V  [libjvm.so+0x14b28e2]  JavaThread::thread_main_inner()+0x242
V  [libjvm.so+0x14b794a]  Thread::call_run()+0xfa
V  [libjvm.so+0x10fc9bf]  thread_native_entry(Thread*)+0x11f
C  [libpthread.so.0+0x63bd]  start_thread+0xfd
```

compiler/intrinsics/string/TestStringLatin1IndexOfChar.java seems fine on my x86_32.

I didn't take a further look at the crash since it's already weekend now.
Thanks.

File: src/hotspot/share/c1/c1_GraphBuilder.cpp [hunk 1]
```diff
@@ -1123,14 +1123,10 @@ void GraphBuilder::stack_op(Bytecodes::Code code) {
 void GraphBuilder::arithmetic_op(ValueType* type, Bytecodes::Code code, ValueStack* state_before) {
   Value y = pop(type);
   Value x = pop(type);
-  // NOTE: strictfp can be queried from current method since we don't
-  // inline methods with differing strictfp bits
-  Value res = new ArithmeticOp(code, x, y, method()->is_strict(), state_before);
+  Value res = new ArithmeticOp(code, x, y,state_before);
```
Comment (iwanowww) [inline src/hotspot/share/c1/c1_GraphBuilder.cpp:1126]: Please, put a space before `state_before`.

**Comment by iwanowww** [comment]
Thanks, Jie.

I think there's `result->is_double_fpu()` check missing in `FpuStackAllocator::handle_op2`:
```
     case lir_mul:
     case lir_div: {
       assert(op2->tmp1_opr()->is_fpu_register(), "strict operations need temporary fpu stack slot");
       insert_free_if_dead(op2->tmp1_opr());
       assert(sim()->stack_size() <= 7, "at least one stack slot must be free");
       // fall-through: continue with the normal handling of lir_mul and lir_div
     }
```

The code should be guarded by `result->is_double_fpu()` since special handling (additional temp operand) is needed only for `mul`/`div` on doubles.

**Comment by DamonFool** [comment]
> Right I just tracked that through before seeing this email. The fmul
> uses the non-tmp version but later all lir_mul are assumed the same.

compiler/c1/Test6855215.java passed on x86_32 with the latest version.

More testing on x86_32 is in progress.
Will let you know once finished.
Thanks.

**Comment by DamonFool** [comment]
> > Right I just tracked that through before seeing this email. The fmul
> > uses the non-tmp version but later all lir_mul are assumed the same.
> 
> compiler/c1/Test6855215.java passed on x86_32 with the latest version.
> 
> More testing on x86_32 is in progress.
> Will let you know once finished.

All the tests passed on our Linux/x86_{64, 32}, no regression.
Thanks.

**Review by iwanowww — APPROVED** [review]

File: src/hotspot/cpu/aarch64/c1_LIRGenerator_aarch64.cpp [hunk 2]
```diff
@@ -423,11 +423,11 @@ void LIRGenerator::do_ArithmeticOp_FPU(ArithmeticOp* x) {
 
   LIR_Opr reg = rlock(x);
   LIR_Opr tmp = LIR_OprFact::illegalOpr;
-  if (x->is_strictfp() && (x->op() == Bytecodes::_dmul || x->op() == Bytecodes::_ddiv)) {
+  if (x->op() == Bytecodes::_dmul || x->op() == Bytecodes::_ddiv) {
     tmp = new_register(T_DOUBLE);
```
Comment (nick-arm) [inline src/hotspot/cpu/aarch64/c1_LIRGenerator_aarch64.cpp:427]: This variable `tmp` doesn't seem to be used. I guess this code originally came from x86 where it's passed through to `arithmetic_op_fpu()`.

**Pull request closed without merging**